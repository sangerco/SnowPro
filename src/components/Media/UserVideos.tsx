import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchVideosByUsername } from "../../redux/slices/mediaSlices";
import { useParams } from "react-router";
import { Button, Dimmer, Header, Loader } from "semantic-ui-react";
import { useNavigate } from "react-router";
import VideoView from "./VideoView";

const UserVideos: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const { username } = useParams();

  useEffect(() => {
    if (username) {
      dispatch(fetchVideosByUsername(username));
    }
  }, [dispatch, username]);

  const media = useSelector((state: RootState) => state.media);
  const videos = media.videos;

  if (media.loading) {
    <Dimmer active>
      <Loader>Loading...</Loader>
    </Dimmer>;
  }

  if (media.error) {
    <Dimmer active>
      <Header as="h1">
        Error! Videos could not be retrieved! {media.error}
      </Header>
      <Button color="red" onClick={() => navigate(-1)}>
        Go Back
      </Button>
    </Dimmer>;
  }

  if (videos) {
    return (
      <div>
        {videos.length > 0
          ? videos.map((video) => <VideoView id={video.id} />)
          : null}
      </div>
    );
  }
};

export default UserVideos;
