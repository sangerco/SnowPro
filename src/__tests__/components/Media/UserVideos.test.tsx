/* eslint-disable testing-library/no-node-access */
import React from "react";
import UserVideos from "../../../components/Media/UserVideos";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test User Videos", () => {
  const store = mockStore({
    auth: {
      data: {
        id: "11",
      },
    },
    media: {
      videos: [
        {
          userId: "11",
          username: "testuser",
          id: "video-1",
          link: "https://www.youtube.com/v=fakevideo1",
          about: "test about video",
          createdAt: "fake created date",
        },
        {
          userId: "11",
          username: "testuser",
          id: "video-2",
          link: "https://www.youtube.com/v=fakevideo2",
          about: "test about video",
          createdAt: "fake created date",
        },
      ],
    },
  });
  it("renders without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserVideos />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UserVideos />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("testing video views", () => {
  it(`show a user's videos`, async () => {
    const username = "testuser";
    const useParamsMock = jest.requireMock("react-router-dom")
      .useParams as jest.Mock;
    useParamsMock.mockReturnValue({ username });

    const store = mockStore({
      auth: {
        data: {
          id: "11",
        },
      },
      media: {
        videos: [
          {
            userId: "11",
            username: "testuser",
            id: "video-1",
            link: "https://www.youtube.com/v=fakevideo1",
            about: "test about video",
            createdAt: "fake created date",
          },
          {
            userId: "11",
            username: "testuser",
            id: "video-2",
            link: "https://www.youtube.com/v=fakevideo2",
            about: "test about video",
            createdAt: "fake created date",
          },
        ],
      },
    });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserVideos />
        </BrowserRouter>
      </Provider>
    );

    // @ts-ignore
    const videos = store.getState((state) => state.media.videos);
    // @ts-ignore
    expect(videos.media.videos.length).toBe(2);
  });
});
