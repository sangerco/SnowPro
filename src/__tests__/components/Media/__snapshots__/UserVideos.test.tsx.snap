// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User Videos matches snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
